### PR TITLE
BUG: unwrap setitem indexer for ArrowExtensionArray

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.0.0.rst` file if fixing a bug or adding a new feature.

Setting values in a pyarrow-backed column with a non-scalar indexer was raising. 

Fixes the following:

```
import pandas as pd

df = pd.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]}, dtype="int64[pyarrow]")

# IndexError: index 1 is out of bounds for axis 0 with size 1
df.loc[[0, 1], "b"] = 0
```

### CODE REVIEW

**Comment by lukemanley** [comment]
The test added in this PR seems to uncover another potential bug. The new test is failing for a separate parameterization:

`pandas/tests/extension/test_numpy.py::TestSetitem::test_setitem_sequence_frame`

However, the failure is only observable when the monkeypatch that is used in the test is applied. Here is a variation of the test to highlight the difference with and without the monkeypatch:

```
import pandas as pd

data = pd.Series([(i,) for i in range(5)]).array

df = pd.DataFrame({"a": data, "b": data})
df.loc[[0, 1], "b"] = [(-1,), (-2,)]
print(df)

# monkeypatch applied in pandas/tests/extension/test_numpy.py
pd.arrays.PandasArray._typ = "extension"

df = pd.DataFrame({"a": data, "b": data})
df.loc[[0, 1], "b"] = [(-1,), (-2,)]
print(df)
```

Without monkeypatch:
```
      a      b
0  (0,)  (-1,)
1  (1,)  (-2,)
2  (2,)   (2,)
3  (3,)   (3,)
4  (4,)   (4,)
```

With monkeypatch:
```
      a     b
0  (0,)    -1
1  (1,)    -2
2  (2,)  (2,)
3  (3,)  (3,)
4  (4,)  (4,)
```

It is not obvious to me how to fix this. I tried a few changes to `_setitem_with_indexer_split_path` but all were unsuccessful. If anyone has ideas, please let me know. Thanks!

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -1669,7 +1669,11 @@ def _unwrap_setitem_indexer(self, indexer):
             #  Should never have length > 2.  Caller is responsible for checking.
             #  Length 1 is reached vis setitem_single_block and setitem_single_column
             #  each of which pass indexer=(pi,)
-            if len(indexer) == 2:
+            if len(indexer) == 1:
```
Comment (phofl) [inline pandas/core/internals/blocks.py:1672]: Not sure, but it might make more sense to handle this in __setitem__ of the ArrowExtensionArray?

Comment (lukemanley) [reply pandas/core/internals/blocks.py:1672]: I was 50/50 on whether to put it here or in ArrowExtensionArray.__setitem__. I ended up leaning towards this as it seems consistent with this method and may be helpful beyond ArrowExtensionArray (e.g. third-party EAs). But happy to move it if setitem is a better place for it.

Comment (phofl) [reply pandas/core/internals/blocks.py:1672]: I am more concerned with our setiteim for Arrow diverging from the numpy setitem call, but not sure if this is correct either.

cc @jbrockmendel Could you weigh in here?

Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:1672]: id want to handle it correctly in the ArrowArray method.  could add a test to the extension base tests for handling this case correctly

Comment (lukemanley) [reply pandas/core/internals/blocks.py:1672]: ok, I will change this to be handled within `ArrowExtensionArray.__setitem__`.  

I guess the typing on `ExtensionArray.__setitem__` and `ArrowExtensionArray.__setitem__` should change as well?

https://github.com/pandas-dev/pandas/blob/14881577a2bd874ee45dde7230cf334b69f9026e/pandas/core/arrays/base.py#L351

https://github.com/pandas-dev/pandas/blob/14881577a2bd874ee45dde7230cf334b69f9026e/pandas/core/arrays/arrow/array.py#L878

Comment (phofl) [reply pandas/core/internals/blocks.py:1672]: Yeah would make sense

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @lukemanley

**Pull request merged**